velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,208 @@
+
+namespace {
+
+int64_t rotateLeft(const int64_t num, const int32_t distance) {
+  auto unsignedShift = ((uint64_t)num) >> -distance;
+  return (num << distance) | unsignedShift;
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t getHashForIntTypes(const T& value) {
+  return rotateLeft(value * 0xC2B2AE3D27D4EB4FL, 31) * 0x9E3779B185EBCA87L;
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t
+getHashForIntTypes(const DecodedVector& vector, vector_size_t row) {
+  auto value = vector.valueAt<T>(row);
+  return getHashForIntTypes(value);
+}
+
+FOLLY_ALWAYS_INLINE int64_t
+getHashForVarCharTypes(const DecodedVector& vector, vector_size_t row) {
+  auto input = vector.valueAt<StringView>(row);
+  int64_t hash = XXH64(input.data(), input.size(), 0);
+  return hash;
+}
+} // namespace
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BIGINT>(vector_size_t row) {
+  return getHashForIntTypes<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::INTEGER>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::SMALLINT>(vector_size_t row) {
+  return getHashForIntTypes<int16_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TINYINT>(vector_size_t row) {
+  return getHashForIntTypes<int8_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DOUBLE>(vector_size_t row) {
+  auto value = vector_->valueAt<double>(row);
+  int64_t u = *reinterpret_cast<int64_t*>(&value);
+  return getHashForIntTypes(u);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DATE>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARCHAR>(vector_size_t row) {
+  return getHashForVarCharTypes(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::VARBINARY>(vector_size_t row) {
+  return getHashForVarCharTypes(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::REAL>(vector_size_t row) {
+  auto value = vector_->valueAt<double>(row);
+  int32_t u = *reinterpret_cast<int32_t*>(&value);
+  return getHashForIntTypes(u);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TIMESTAMP>(vector_size_t row) {
+  return getHashForIntTypes<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BOOLEAN>(vector_size_t row) {
+  auto value = vector_->valueAt<bool>(row);
+  return value ? 1231 : 1237;
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::ARRAY>(vector_size_t row) {
+  int64_t hash = 0;
+  auto baseArray = vector_->base()->as<velox::ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = vector_->indices();
+
+  auto size = rawSizes[indices[row]];
```

Comment:
rawSizes and rawOffsets are used only once, hence, it would be simpler to use sizeAt and offsetAt methods:

```
auto index = vector->indexOf(row);
auto size =  baseArray->sizeAt(index);
auto offset =  baseArray->offsetAt(index);
```